File: JavaScript/maxProduct.js
/**
 * https://leetcode.com/problems/maximum-product-subarray/
 * @param {number[]} nums
 * @return {number}
 * Time O(n) Space O(1)
 * 
 * Unlike add, max product need to rememeber both max value and min value in case of negtive number.
 * And in case one negative number multiplate another negtive number and become positive, we need to 
 * compare both of them before compare to current number.
 */
var maxProduct = function(nums) {
	var maxPre = nums[0],
		minPre = nums[0],
		maxSoFar = nums[0],
		maxHere, minHere;
	for (var i = 1, len = nums.length; i < len; i++) {
		maxHere = Math.max(Math.max(maxPre * nums[i], minPre * nums[i]), nums[i]);
		minHere = Math.min(Math.min(minPre * nums[i], maxPre * nums[i]), nums[i]);
		maxSoFar = Math.max(maxHere, maxSoFar);
		maxPre = maxHere;
		minPre = minHere;

	}
	return maxSoFar;
};

File: Python/maxProduct.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        maxPre = minPre = maxNew = nums[0]
        for i in xrange(1,len(nums)):
            maxHere = max(max(maxPre*nums[i], minPre*nums[i]), nums[i])
            minHere = min(min(maxPre*nums[i], minPre*nums[i]), nums[i])
            maxNew = max(maxHere, maxNew) #compare current max and storaged max value
            maxPre = maxHere
            minPre = minHere
        return maxNew
            
